### PR TITLE
support pulsar schema

### PR DESCRIPTION
**Pull request by jianyun8023**
Support Pulsar schema

- add `PulsarContextAware` interface.

### CODE REVIEW

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/PulsarPartitionReader.java [hunk 1]
```diff
@@ -63,33 +61,33 @@ public PulsarPartitionSplit getSplit() {
         return split;
     }
 
-    public Iterator<Message<?>> nextBatch() throws PulsarClientException {
+    public Iterator<Message<V>> nextBatch() throws PulsarClientException {
         if (consumer.hasMessageAvailable()) {
-            Messages<byte[]> messages = consumer.batchReceive();
-            Iterator<Message<byte[]>> messageIterator = messages.iterator();
+            Messages<V> messages = consumer.batchReceive();
```
Comment (codelipenghui) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/PulsarPartitionReader.java:66]: Any reason for using `batchReceive()` here? The current implementation for the `batchReceive()` can’t improve performance for message reading. And the default behavior for the `BatchReceivePolicy` is to wait at most 100ms if there are not enough messages for a batch(default is 1MB).

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/FlinkSchema.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            SchemaInfo schemaInfo,
+            SerializationSchema<T> serializer,
+            DeserializationSchema<T> deserializer) {
+        this.schemaInfo = schemaInfo;
+        this.serializer = serializer;
+        this.deserializer = deserializer;
+    }
+
+    @Override
+    public void validate(byte[] message) {}
```
Comment (syhily) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/FlinkSchema.java:53]: Add the reason why this is a blank method.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/FlinkSchema.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public byte[] encode(T t) {
+        if (serializer == null) {
+            throw new UnsupportedOperationException();
```
Comment (syhily) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/FlinkSchema.java:58]: Add a more clear exception message.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/FlinkSchema.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        }
+        return serializer.serialize(t);
+    }
+
+    @Override
+    public T decode(byte[] bytes) {
+        if (deserializer == null) {
+            throw new UnsupportedOperationException();
```
Comment (syhily) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/FlinkSchema.java:66]: Same here. Add a more clear exception message.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/PulsarPartitionReader.java (hunk 1 above)
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/PulsarPartitionReader.java:66]: No reason for `batchReceive`, we would remove it.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/PulsarContextAware.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * An interface for {@link MessageDeserializer dynamicPulsarSerializationSchema} that need
+ * information about the context where the Pulsar Producer is running along with information about
+ * the available partitions.
+ *
+ * <p>You only need to override the methods for the information that you need. However, {@link
+ * #getTargetTopic(Object)} is required because it is used to determine the available partitions.
+ */
+@PublicEvolving
+public interface PulsarContextAware<T> extends ResultTypeQueryable<T> {
```
Comment (nlu90) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/PulsarContextAware.java:37]: This class naming is strange. Is this a convention in Flink?

If no, can we rename it to `PulsarContext`?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/deserializer/MessageDeserializer.java [hunk 6]
```diff
@@ -51,8 +55,13 @@
...
             @Override
-            public void deserialize(Message<?> message, Collector<V> collector) throws IOException {
-                valueDeserializer.deserialize(message.getData(), collector);
+            public Schema<V> getSchema() {
+                return new FlinkSchema<>(Schema.BYTES.getSchemaInfo(), null, valueDeserializer);
+            }
+
+            @Override
+            public void deserialize(Message<V> message, Collector<V> collector) throws IOException {
+                collector.collect(message.getValue());
```
Comment (nlu90) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/deserializer/MessageDeserializer.java:64]: should the `valueDeserializer` deserialize the message value first?

**Review by nlu90 — COMMENTED** [review]
Another general question: How is the mapping of Flink Schema and Pulsar Schema implemented.

**Pull request closed without merging**